concordium-consensus/src/Concordium/Scheduler.hs

```diff
@@ -1280,33 +1294,41 @@ handleChainUpdate WithMetadata{wmdData = ui@UpdateInstruction{..}, ..} = do
+            Just ai -> checkSigAndEnqueue $ UVFoundationAccount ai
             Nothing -> return (TxInvalid (UnknownAccount u))
-          MintDistributionUpdatePayload u -> checkSigAndUpdate $ UVMintDistribution u
-          TransactionFeeDistributionUpdatePayload u -> checkSigAndUpdate $ UVTransactionFeeDistribution u
-          GASRewardsUpdatePayload u -> checkSigAndUpdate $ UVGASRewards u
-          BakerStakeThresholdUpdatePayload u -> checkSigAndUpdate $ UVBakerStakeThreshold u
-          AddAnonymityRevokerUpdatePayload u -> checkSigAndUpdate $ UVAddAnonymityRevoker u
-          AddIdentityProviderUpdatePayload u -> checkSigAndUpdate $ UVAddIdentityProvider u
-          RootUpdatePayload (RootKeysRootUpdate u) -> checkSigAndUpdate $ UVRootKeys u
-          RootUpdatePayload (Level1KeysRootUpdate u) -> checkSigAndUpdate $ UVLevel1Keys u
-          RootUpdatePayload (Level2KeysRootUpdate u) -> checkSigAndUpdate $ UVLevel2Keys u
-          Level1UpdatePayload (Level1KeysLevel1Update u) -> checkSigAndUpdate $ UVLevel1Keys u
-          Level1UpdatePayload (Level2KeysLevel1Update u) -> checkSigAndUpdate $ UVLevel2Keys u
-
+          MintDistributionUpdatePayload u -> checkSigAndEnqueue $ UVMintDistribution u
+          TransactionFeeDistributionUpdatePayload u -> checkSigAndEnqueue $ UVTransactionFeeDistribution u
+          GASRewardsUpdatePayload u -> checkSigAndEnqueue $ UVGASRewards u
+          BakerStakeThresholdUpdatePayload u -> checkSigAndEnqueue $ UVBakerStakeThreshold u
+          AddAnonymityRevokerUpdatePayload u -> checkSigAndEnqueue $ UVAddAnonymityRevoker u
+          AddIdentityProviderUpdatePayload u -> checkSigAndEnqueue $ UVAddIdentityProvider u
+          RootUpdatePayload (RootKeysRootUpdate u) -> checkSigAndEnqueue $ UVRootKeys u
+          RootUpdatePayload (Level1KeysRootUpdate u) -> checkSigAndEnqueue $ UVLevel1Keys u
+          RootUpdatePayload (Level2KeysRootUpdate u) -> checkSigAndEnqueue $ UVLevel2Keys u
+          Level1UpdatePayload (Level1KeysLevel1Update u) -> checkSigAndEnqueue $ UVLevel1Keys u
+          Level1UpdatePayload (Level2KeysLevel1Update u) -> checkSigAndEnqueue $ UVLevel2Keys u
   where
-    checkSigAndUpdate change = do
-      -- Check that the signatures use the appropriate keys and are valid.
-      keyCollection <- getUpdateKeyCollection
-      if checkAuthorizedUpdate keyCollection ui then do
-        enqueueUpdate (updateEffectiveTime uiHeader) change
-        tsIndex <- bumpTransactionIndex
-        return $ TxValid TransactionSummary {
+    checkSigAndEnqueue change = do
+      cache <- getTransactionVerificationCache
+      case HM.lookup wmdHash (cache ^. chainUpdates) of
+        Just (ChainUpdateVerificationResult keysHash _) -> do
+          currentKeys <- getUpdateKeyCollection
+          if getHash currentKeys == keysHash then do
+            update change
+          else return $ TxInvalid IncorrectSignature -- keys changed the signature is no longer valid
```

Comment:
I don't think this is correct since keys can change and the signature can remain valid (because keys that did not sign could have changed, which does not matter).